mqtt/mqtt-bridge/src/client.rs

```diff
@@ -184,6 +186,36 @@ impl BridgeIoSource {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct InFlightPublishHandle {
+    publish_handle: PublishHandle,
+    permits: Arc<Semaphore>,
+}
+
+impl InFlightPublishHandle {
+    fn new(publish_handle: PublishHandle, permits: Arc<Semaphore>) -> Self {
+        Self {
+            publish_handle,
+            permits,
+        }
+    }
+
+    pub async fn publish(&self, publication: Publication) {
+        let mut publish_handle = self.publish_handle.clone();
+        let permits = self.permits.clone();
+
+        let fut = async move {
+            let permit = permits.acquire().await;
+            if let Err(e) = publish_handle.publish(publication).await {
+                error!(message = "failed to publish", err = %e);
+            }
+            drop(permit)
```

Comment:
I guess you don't need to drop that manually, because the scope ends right on the next line